askama_shared/src/generator.rs

```diff
@@ -822,36 +848,24 @@ impl<'a, S: std::hash::BuildHasher> Generator<'a, S> {
         let mut expr_buf = Buffer::new(0);
         self.visit_expr(&mut expr_buf, val)?;
 
+        let (shadowed, declared) = self.shadows_or_declares_variable(var);
+        if shadowed {
+            // Need to flush the buffer if the variable is being shadowed,
+            // to ensure the old variable is used.
+            self.write_buf_writable(buf)?;
+        }
+        if declared {
+            buf.write("let ");
+        }
```

Comment:
Would it make sense to inline these effects into `shadows_or_declares_variable()` (I suppose it would need yet another name)? Splitting it this way doesn't make for a great abstraction.